File: leetCode/python/p23_MergeKSortedLists/lc23.py
#!/usr/bin/python3

class ListNode:
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next

class Solution:
	'''
	You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

	Merge all the linked-lists into one sorted linked-list and return it.
	'''
	def __init__(self):
		self.head = None

	def append_node(self, val):
		new_node = ListNode(val)
		# check if head already exists
		if self.head:
			current = self.head
			# go to the end of linked list
			while current.next:
				current = current.next
			current.next = new_node
		else:
			self.head = new_node

	def append_nodes_from_list(self, val_list: list):
		for i in range(len(val_list)):
			self.append_node(val_list[i])

	def print_linked_list(self):
		current = self.head
		
		# if list is empty print []	
		if current == None:
			print('[]', end = ' ')
	
		while current:
			print(current.val, end = ' ')
			current = current.next
		print()

def mergeTwoLists(list1: ListNode, list2: ListNode) -> ListNode:				
	head = temp = ListNode()

	while list1 or list2:	
		# if one of the lists is empty
		if list1 and not list2:
			while list1:
				temp.next = ListNode(list1.val)
				list1 = list1.next	
			break
		if list2 and not list1:	
			while list2:
				temp.next = ListNode(list2.val)
				list2 = list2.next	
			break

		if list1.val <= list2.val:
			temp.next = ListNode(list1.val)
			list1 = list1.next
		else:
			temp.next = ListNode(list2.val)
			list2 = list2.next
	
		# advance the pointer
		temp = temp.next 

	return head.next

def mergeKLists(lists):

	sol = Solution()
	temp = Solution()
	list_size = len(lists)
	
	# check if list is empty
	if list_size == 0:	
		return sol
	# check if all the rows in the list are empty
	else: 						
		all_rows_empty = True
		for row in lists:
			if len(row) != 0:
				all_rows_empty = False
		
		if all_rows_empty:
			return sol
		
	sol.append_nodes_from_list(lists[0])

	if list_size > 1:
		for i in range(1, list_size):
			temp.append_nodes_from_list(lists[i])
			sol.head = mergeTwoLists(sol.head, temp.head)
			temp.head = None

	return sol
		
sol = Solution()

# Test 1

element_lists1 = [[1, 4, 5],[1, 3, 4],[2, 6]]

sol = mergeKLists(element_lists1)
print('Test 1:')
sol.print_linked_list()
sol.head = None

# Test 2

element_lists2 = []

sol = mergeKLists(element_lists2)
print('Test 2:')
sol.print_linked_list()
sol.head = None

# Test 3

element_lists3 = [[]]

sol = mergeKLists(element_lists3)
print('Test 3:')
sol.print_linked_list()
sol.head = None

File: leetCode/c/p23_MergeKSortedLists/lc23.c
/* You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

Merge all the linked-lists into one sorted linked-list and return it. */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "../../../mylib/mylib.h"

struct ListNodeInt* mergeTwoLists(struct ListNodeInt* list1, struct ListNodeInt* list2) {
	
	struct ListNodeInt *pnt;	
	listInitializeInt(&pnt);

	struct ListNodeInt *head = pnt;

	// Check if both lists are empty
	if(list1->next == NULL && list1->val == -1 && list2->next == NULL && list2->val == -1) 
		return head;

	while ((list1->next != NULL) || (list2->next != NULL)) {
	
		// If end of list2 is reached, but not the end of list1	
		if((list1->next != NULL) && (list2->next == NULL)) {
			
			while(list1->next != NULL) {
				listPushInt(pnt, list1->val);
				list1 = list1->next;
			}

			break;
		}

		// If end of list1 is reached, but not the end of list2
		if((list1->next == NULL) && (list2->next != NULL)) {
				
			while(list2->next != NULL) {
				listPushInt(pnt, list2->val);
				list2 = list2->next;
			}

			break;
		}
	
		if(list1->val <= list2->val) {
			listPushInt(pnt, list1->val);
			list1 = list1->next;
		}
		else {	
			listPushInt(pnt, list2->val);
			list2 = list2->next;
		}

	}
	
	// Add the last list elements
	if(list1->val <= list2->val) {
		listPushInt(pnt, list1->val);
		listPushInt(pnt, list2->val);
	}
	else{
		listPushInt(pnt, list2->val);
		listPushInt(pnt, list1->val);
	}

	return head;
}

void mergeKLists(struct ListNodeInt** lists, int listsSize) {

	struct ListNodeInt *sol;
	struct ListNodeInt *emptyList;

	listInitializeInt(&emptyList);	
		
	/* Check if list is empty */
	if(listsSize == 0) {
		listPrintInt(emptyList);
		return;
	}
	/* Check if all the rows in the list are empty */	
	else {
		bool allRowsEmpty = true;
		for(int i=0; i<listsSize; i++) {
			if(lists[i]->val != -1 && lists[i]->next != NULL)
				allRowsEmpty = false;	
		}
	
		if(allRowsEmpty) {
			listPrintInt(emptyList);
			return;
			}
	}	

	sol = lists[0];

	for(int i=1; i<listsSize; i++) {
		sol = mergeTwoLists(sol, lists[i]);		
	}

	listPrintInt(sol);
}

int main(void) {

	struct ListNodeInt *sol;

	/* Test */

	struct ListNodeInt **lists1;
	struct ListNodeInt *list1, *list2, *list3;

	listInitializeInt(&list1);	
	listInitializeInt(&list2);
	listInitializeInt(&list3);

	int list1Array[3] = {1, 4, 5};	
	int list1Size = sizeof(list1Array)/sizeof(int);
	int list2Array[3] = {1, 3, 4};
	int list2Size = sizeof(list2Array)/sizeof(int);
	int list3Array[2] = {2, 6};
	int list3Size = sizeof(list3Array)/sizeof(int);

	listInsertArrayInt(list1, list1Array, list1Size);
	listInsertArrayInt(list2, list2Array, list2Size);
	listInsertArrayInt(list3, list3Array, list3Size);

	lists1[0] = list1;	
	lists1[1] = list2;
	lists1[2] = list3;
	
	printf("Test 1:\n");
	mergeKLists(lists1, 3); 

	/* Test 2 */

	printf("\nTest 2:\n");
	/* Simulate empty list */
	mergeKLists(lists1, 0); 	
	
	return EXIT_SUCCESS;
}
